gk/main.c

```diff
@@ -1418,6 +1420,279 @@ update_ip_hop_count(struct gatekeeper_if *iface, struct ipacket *packet,
 	return 0;
 }
 
+static void
+lookup_fib_bulk(struct gk_lpm *ltbl, struct ip_flow **flows,
+	int num_flows, struct gk_fib *fibs[])
+{
+	int i;
+	/* The batch size for IPv4 LPM table lookup. */
+	const uint8_t FWDSTEP = 4;
+	const uint32_t default_nh = 0xFFFFFF;
+	int k = RTE_ALIGN_FLOOR(num_flows, FWDSTEP);
+
+	for (i = 0; i != k; i += FWDSTEP) {
+		int j;
+		__m128i dip = _mm_set_epi32(flows[i + 3]->f.v4.dst.s_addr,
+			flows[i + 2]->f.v4.dst.s_addr,
+			flows[i + 1]->f.v4.dst.s_addr,
+			flows[i]->f.v4.dst.s_addr);
+		rte_xmm_t dst;
+
+		rte_lpm_lookupx4(ltbl->lpm, dip, dst.u32, default_nh);
+
+		for (j = 0; j < FWDSTEP; j++) {
+			if (dst.u32[j] != default_nh)
+				fibs[i + j] = &ltbl->fib_tbl[dst.u32[j]];
+			else
+				fibs[i + j] = NULL;
+		}
+	}
+
+	for (i = k; i < num_flows; i++)
+		fibs[i] = look_up_fib(ltbl, flows[i]);
+}
+
+static void
+lookup_fib6_bulk(struct gk_lpm *ltbl, struct ip_flow **flows,
+	int num_flows, struct gk_fib *fibs[])
+{
+	int i;
+	uint8_t dst_ip[num_flows][RTE_LPM6_IPV6_ADDR_SIZE];
+	int32_t hop[num_flows];
+
+	for (i = 0; i < num_flows; i++) {
+		memcpy(&dst_ip[i][0], flows[i]->f.v6.dst.s6_addr,
+			sizeof(dst_ip[i]));
+	}
+
+	rte_lpm6_lookup_bulk_func(ltbl->lpm6, dst_ip, hop, num_flows);
+
+	for (i = 0; i < num_flows; i++) {
+		if (hop[i] != -1)
+			fibs[i] = &ltbl->fib_tbl6[hop[i]];
+		else
+			fibs[i] = NULL;
+	}
+}
+
+static struct flow_entry *
+lookup_fe_from_lpm(struct ipacket *packet, uint32_t ip_flow_hash_val,
```

Comment:
I'm aware of the extra work, but putting `lookup_fe_from_lpm()` and  `process_flow_entry()` into a first (new) patch would help the review of this patch, namely, **gk: batch flow and LPM lookups**.

Reply:
Good points! Done.